hummingbot/client/command/gateway_command.py

```diff
@@ -63,6 +68,104 @@ async def _generate_certs(self,  # type: HummingbotApplication
+
+        if len(listdir(gateway_conf_path)) == 1:
+            self._notify("Initiating Gateway configurations from sample files...")
+            shutil.copytree(path.join(gateway_conf_path, "samples"), gateway_conf_path, dirs_exist_ok=True)
+            self.app.clear_input()
+            self.placeholder_mode = True
+            self.app.hide_input = True
+            # prompt for questions about infura key
+            use_infura = await self.app.prompt(prompt="Would you like to connect to Ethereum network using Infura? (Yes/No) >>> ")
+            yaml_parser = ruamel.yaml.YAML()
+            ethereum_conf_path = path.join(gateway_conf_path, "ethereum.yml")
+            if use_infura is not None and use_infura in ["Y", "y", "Yes", "yes"]:
+                infura_key = await self.app.prompt(prompt="Enter your Infura API key >>> ")
+                try:
+                    with open(ethereum_conf_path) as stream:
+                        data = yaml_parser.load(stream) or {}
+                        for key in data["networks"]:
+                            data["networks"][key]["nodeApiKey"] = infura_key
+                        with open(ethereum_conf_path, "w+") as outfile:
+                            yaml_parser.dump(data, outfile)
+                except Exception as e:
+                    self.logger().error("Error writing configs: %s" % (str(e),), exc_info=True)
+            else:
+                node_rpc = await self.app.prompt(prompt="Enter node rpc url to use to connect to Ethereum mainnet  >>> ")
+                try:
+                    with open(ethereum_conf_path) as stream:
+                        data = yaml_parser.load(stream) or {}
+                        data["networks"]["mainnet"]["nodeURL"] = node_rpc
+                        with open(ethereum_conf_path, "w+") as outfile:
+                            yaml_parser.dump(data, outfile)
+                except Exception:
+                    self.logger().error("Error updating Ethereum mainnet rpc url.")
+            self.placeholder_mode = False
+            self.app.hide_input = False
+            self.app.change_prompt(prompt=">>> ")
+
+        # remove existing container(s)
+        try:
+            old_container = self._docker_client.containers(all=True,
+                                                           filters={"name": gateway_container_name})
+            for container in old_container:
+                self._notify(f"Removing existing gateway container with id {container['Id']}...")
+                self._docker_client.remove_container(container["Id"], force=True)
+        except Exception:
+            pass  # silently ignore exception
+
+        await self._generate_certs()  # create cert if not available
+        self._notify("Pulling Gateway docker image...")
+        await asyncio.sleep(0.5)
+        try:
+            pull_logs = iter(self._docker_client.pull(gateway_docker_name, tag="gateway-v2", stream=True, decode=True))
```

Comment:
The iterator here is synchronous, which causes the `prompt-toolkit` UI library to be unresponsive during the image pull.

There are two options here:

 1. Wrap this in a thread via `run_in_executor()`, and then use `call_soon_threadsafe()` from the child thread to schedule the message back to the main thread. You can then wrap this all again into an async iterator, s.t. you can do `async for message in thread_pull_wrapper.pull(...): ...`.
 2. Use `aiokafka` instead. They also have a stream option.